### DISCUSSION TITLE
Really need help - spending  days trying  this autoconnect functionality

### DISCUSSION THREAD

**Discussion by rowifi** [Hieromon/AutoConnect#491] [How can I?] [answered]
I want to do something which I think should be a really standard procedure, but I cannot get it to work and I've tried many variations.
Using ESP32 with platformIO.

I have two buttons on the ESP32 which I would use to Remove the Wifi credentials ( set them to 0,0 ) on a power up cycle for a fresh start without wifi capability. And another to start the SoftAP.

From a power up, the app should try to connect for a minute or so ( if it has any credentials other than 0,0 ) and then give up, decide there is no wifi and continue to run the program ( no softap ). Any Wifi functionality will of course be prohibited.
The softAP should never appear without the user pressing the button because the network may be down and we would want it to reconnect automatically  (just like my phone or tablet does ).
As long as there are valid credentials, then the device should periodically try to connect.

If the user decides they simply don't want the wifi functionality they would never use the button to enter their credentials, and the program would begin rapidly without any wifi capability.
I've spent days trying to solve this.. I've also continued to get an error:
 [E][Preferences.cpp:49] begin(): nvs_open failed: NOT_FOUND

Which I also cannot resolve.
Any help appreciated.

**Comment by tktf50** [comment, marked as answer]
Hi @rowifi,

If you are still looking for help, it would be a good idea to post the relevant portions of your code. It's hard to tell from your original post whether you are having problems with specific parts of the code or the entire program.

However, I use a similar approach in one of my programs - so this may help you. I use a button to put the MCU into "configuration mode" at startup time. If the button is not pressed, the program runs in "normal mode" and attempts to connect to the WiFi network periodically.

```
WebServer   webServer;              // Web server
AutoConnect autoPortal(webServer);  // Control of WiFi connection and configuration portal

static bool doConfiguration = false;
static unsigned long lastConnectionAttempt = 50000L;          // Last time we attempted to connect to the network.

const uint32_t MILLIS_PER_SECOND = 1000;                      // Number of milliseconds in a second
const uint32_t MILLIS_PER_MINUTE = 60 * MILLIS_PER_SECOND;    // Number of milliseconds in a minute
const long CONNECT_RETRY_INTERVAL = 10 * MILLIS_PER_MINUTE;   // Number of milliseconds to wait between attempts
                                                              //  to connect to the local network.
const unsigned long CONNECT_TIMEOUT = 20 * MILLIS_PER_SECOND; // Number of milliseconds before network
                                                              //  connection attempt times out.

void setup() {
  // Code here sets 'doConfiguration' flag to 'true' if the button is pressed
  // and also starts the AutoConnect portal in that case.
}

void loop() {
  if (doConfiguration) {
    autoPortal.handleClient();
  }
  else {
    // Periodic check for WiFi connection
    if (millis() - lastConnectionAttempt > CONNECT_RETRY_INTERVAL) {
      lastConnectionAttempt = millis();
      // If we are not connected to the LAN, try to connect.
      if (!WiFi.isConnected()) {
        bool bOk;
        lastConnectionAttempt = millis();

        WiFi.disconnect();
        portalConfig.autoRise = false;				// Prevent AutoConnect portal from starting
                                                                //  if WiFi connection cannot be established
        portalConfig.autoReconnect = true;			// Try to connect with any AP for which
                                                                //  you have configured credentials using AutoConnect
        portalConfig.principle = AC_PRINCIPLE_RSSI;             // When multiple AP's available, give preference to
                                                                //  AP with strongest signal first
        bOk = autoPortal.config(portalConfig);
        Serial.printf_P(PSTR("autoPortal.config = %d\n"), bOk);
        // Try to establish a network connection using the credentials stored by AutoConnect library.
        bOk = autoPortal.begin(nullptr, nullptr, CONNECT_TIMEOUT);
        Serial.printf_P(PSTR("autoPortal.begin %d: %s\n"), bOk, WiFi.localIP().toString().c_str());
      }
    }

    // Here you can add your regular loop code, and perform network operations based on
    // result of WiFi.isConnected()
  }
}
```
I don't bother to check if any credentials have been configured before attempting the WiFi connection (because my app does not function well without them) but I believe you could do that with the [AutoConnectCredentials.entries()](https://hieromon.github.io/AutoConnect/credit.html#entries) method. Check out Hieromon's documentation.

Hope that helps.

**Reply by Hieromon** [reply]
@TKTF50 Thank you for your response. I think your approach is versatile and transferable and will be helpful for other sketches.

@rowifi The determination of whether or not the credentials that match the nearby AP are stored is automatically determined internally by AutoConnect::handleClient when autoReconnect and reconnectInterval are used together. TKTF50 suggests the implementation as follows:

```cpp
void loop() {
  if (doConfiguration) {
    autoPortal.handleClient();
  }
```

His implementation is even more fine-grained, with a structure that enables autoReconnect only when WiFi is disconnected. It seems to me that it satisfies your requirements.

You stuck phenomenon:`[E][Preferences.cpp:49] begin(): nvs_open failed: NOT_FOUND` is not an abnormal case for AutoConnect. Nor is it an impediment to your request.
AutoConnect saves credentials to Preferences, but arduino core triggers such an error when none of the credentials are saved in the first place. AutoConnect has an implementation to automatically recover from that situation.
